src/ConnectedMode/ServerSentEvents/ServerSentEventSessionManager.cs

```diff
@@ -38,75 +41,153 @@ internal interface IServerSentEventSessionManager : IDisposable
-        private readonly IThreadHandling threadHandling;
-        private bool disposed;
-        private CancellationTokenSource sessionTokenSource;
+        private readonly ISSESessionFactory sseSessionFactory;
+
+        private ISSESession currentSession;
 
         [ImportingConstructor]
-        public ServerSentEventSessionManager(
-            IActiveSolutionBoundTracker activeSolutionBoundTracker,
-            ISonarQubeService sonarQubeClient,
-            IServerSentEventPump eventPump,
-            IThreadHandling threadHandling)
+        public ServerSentEventSessionManager(IActiveSolutionBoundTracker activeSolutionBoundTracker, ISSESessionFactory sseSessionFactory)
         {
             this.activeSolutionBoundTracker = activeSolutionBoundTracker;
-            this.sonarQubeClient = sonarQubeClient;
-            this.eventPump = eventPump;
-            this.threadHandling = threadHandling;
+            this.sseSessionFactory = sseSessionFactory;
+
             activeSolutionBoundTracker.SolutionBindingChanged += OnSolutionChanged;
         }
 
         public void Dispose()
         {
-            if (disposed)
-            {
-                return;
-            }
-
             activeSolutionBoundTracker.SolutionBindingChanged -= OnSolutionChanged;
-            sessionTokenSource?.Cancel();
-            eventPump.Dispose();
-            disposed = true;
+            currentSession?.Dispose();
         }
 
         private void OnSolutionChanged(object sender, ActiveSolutionBindingEventArgs activeSolutionBindingEventArgs)
         {
             var bindingConfiguration = activeSolutionBindingEventArgs.Configuration;
             var isInConnectedMode = !bindingConfiguration.Equals(BindingConfiguration.Standalone);
 
-            EndSession();
+            currentSession?.Dispose();
 
             if (!isInConnectedMode)
             {
                 return;
             }
 
-            InitializeSessionAsync(bindingConfiguration.Project.ProjectKey).Forget();
+            currentSession = sseSessionFactory.Create(bindingConfiguration.Project.ProjectKey);
+
+            currentSession.PumpAllAsync().Forget();
+        }
+    }
+    
+    internal interface ISSESessionFactory
+    {
+        ISSESession Create(string projectKey);
+    }
+
+    internal interface ISSESession : IDisposable
+    {
+        Task PumpAllAsync();
+    }
+
+    [Export(typeof(ISSESessionFactory))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    internal class SSESessionFactory : ISSESessionFactory
+    {
+        private readonly ISonarQubeService sonarQubeClient;
+        private readonly IIssueChangedServerEventSourcePublisher issueChangedServerEventSourcePublisher;
+        private readonly ITaintServerEventSourcePublisher taintServerEventSourcePublisher;
+        private readonly IThreadHandling threadHandling;
+
+        [ImportingConstructor]
+        public SSESessionFactory(ISonarQubeService sonarQubeClient,
```

Comment:
I'm thinking something nicer can be done here in terms of duplication but that's the gist.